test/mpi/test_mpi_cuda.c

```diff
@@ -0,0 +1,668 @@
+    CUDA_CALL(cuMemAddressFree(alloc_mem->ptr, alloc_mem->size));
+    CUDA_CALL(cuMemRelease((CUmemGenericAllocationHandle)alloc_mem->obj));
+}
+
+#if HAVE_CUDA_FABRIC
+static alloc_mem_t alloc_vmm_fabric(size_t size)
+{
+    return alloc_vmm_type(size, CU_MEM_HANDLE_TYPE_FABRIC);
+}
+
+static alloc_mem_t alloc_mempool(size_t size)
+{
+    CUmemPoolProps pool_props = {};
+    alloc_mem_t alloc_mem     = {};
+    CUdevice cu_dev;
+    CUmemoryPool mpool;
+    CUmemAccessDesc map_desc;
+    CUdeviceptr ptr;
+
+    CUDA_CALL(cuCtxGetDevice(&cu_dev));
+
+    pool_props.allocType     = CU_MEM_ALLOCATION_TYPE_PINNED;
+    pool_props.location.id   = (int)cu_dev;
+    pool_props.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+    pool_props.handleTypes   = CU_MEM_HANDLE_TYPE_FABRIC;
+    pool_props.maxSize       = size;
+    CUDA_CALL(cuMemPoolCreate(&mpool, &pool_props));
+
+    map_desc.flags    = CU_MEM_ACCESS_FLAGS_PROT_READWRITE;
+    map_desc.location = pool_props.location;
+    CUDA_CALL(cuMemPoolSetAccess(mpool, &map_desc, 1));
+    CUDA_CALL(cuMemAllocFromPoolAsync(&ptr, size, mpool, NULL));
+    CUDA_CALL(cuStreamSynchronize(NULL));
+
+    alloc_mem.ptr = ptr;
+    alloc_mem.obj = (void*)mpool;
+
+    return alloc_mem;
+}
+
+void free_mempool(alloc_mem_t *alloc_mem)
+{
+    CUDA_CALL(cuMemFree(alloc_mem->ptr));
+    CUDA_CALL(cuMemPoolDestroy((CUmemoryPool)alloc_mem->obj));
+}
+#endif
+
+allocator_t allocators[] = {{"cuMemAlloc", alloc_mem_alloc, free_mem_alloc},
+                            {"cuMemAllocManaged", alloc_mem_alloc_managed,
+                             free_mem_alloc},
+                            {"VMM", alloc_vmm, free_vmm},
+#if HAVE_CUDA_FABRIC
+                            {"mempool", alloc_mempool, free_mempool},
+                            {"VMM_Fabric", alloc_vmm_fabric, free_vmm}
+#endif
+};
+
+void cuda_memcpy(void *dest, const void *src, size_t n)
+{
+    CUDA_CALL(cuMemcpy((CUdeviceptr)dest, (CUdeviceptr)src, n));
+    CUDA_CALL(cuCtxSynchronize());
+}
+
+static int check_result(CUdeviceptr dptr, const void *gold_data, size_t size)
+{
+    void *h_buffer;
+    int ret;
+
+    h_buffer = malloc(size);
+    if (h_buffer == NULL) {
+        fprintf(stderr, "failed to allocate host memory\n");
+        return -1;
+    }
+
+    cuda_memcpy(h_buffer, (void*)dptr, size);
+    ret = memcmp(h_buffer, gold_data, size);
+    free(h_buffer);
+    return ret;
+}
+
+static int test_alloc_prim_send_prim(const allocator_t *allocator_send,
+                                     const allocator_t *allocator_recv,
+                                     const void *gold_data, size_t size,
+                                     CUdevice cu_dev)
+{
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    int res;
+
+    retain_and_push_primary_context(cu_dev);
+
+    alloc_mem_send = allocator_send->alloc(size);
+    alloc_mem_recv = allocator_recv->alloc(size);
+
+    cuda_memcpy((void*)alloc_mem_send.ptr, gold_data, size);
```

Comment:
maybe it is worth to memset recv buffer to zeros or something

Reply:
Done.